domain/user/service/service.go

```diff
@@ -188,35 +258,53 @@ func (s *Service) AddUserWithPassword(ctx context.Context, user user.User, passw
 // - usererrors.AlreadyExists: If a user with the supplied name already exists.
 // - usererrors.UserCreatorNotFound: If a creator has been supplied for the user
 // and the creator does not exist.
-func (s *Service) AddUserWithActivationKey(ctx context.Context, user user.User) ([]byte, error) {
-	if err := ValidateUsername(user.Name); err != nil {
-		return nil, fmt.Errorf("username %q with activation key: %w", user.Name, err)
+func (s *Service) AddUserWithActivationKey(ctx context.Context, usr user.User, creatorUUID user.UUID) ([]byte, error) {
+	// Validate user name and creator name
+	if err := ValidateUsername(usr.Name); err != nil {
+		return nil, fmt.Errorf("username %q with activation key: %w", usr.Name, err)
+	}
+
+	// Validate creator UUID
+	if err := creatorUUID.Validate(); err != nil {
+		return nil, fmt.Errorf("creator uuid %q with activation key: %w", creatorUUID, err)
 	}
 
+	// Generate an activation key for the user.
 	activationKey, err := generateActivationKey()
 	if err != nil {
-		return nil, fmt.Errorf("generating activation key for user %q: %w", user.Name, err)
+		return nil, fmt.Errorf("generating activation key for user %q: %w", usr.Name, err)
+	}
+
+	// Get creator UUID for creatorUsername
+	creatorUser, err := s.GetUser(ctx, creatorUUID)
+	if errors.Is(err, usererrors.NotFound) {
+		return nil, fmt.Errorf("%w %q", usererrors.UserCreatorUUIDNotFound, creatorUUID)
+	} else if err != nil {
+		return nil, fmt.Errorf("getting creator user uuid %q: %w", creatorUUID, err)
+	}
+
+	// Generate a UUID for the user.
+	uuid, err := user.NewUUID()
+	if err != nil {
+		return nil, fmt.Errorf("adding user %q with activation key, generating uuid: %w", usr.Name, err)
 	}
 
-	if err = s.st.AddUserWithActivationKey(ctx, user, activationKey); err != nil {
-		return nil, fmt.Errorf("adding user %q with activation key: %w", user.Name, err)
+	if err = s.st.AddUserWithActivationKey(ctx, uuid, usr, creatorUser.UUID, activationKey); err != nil {
+		return nil, fmt.Errorf("adding user %q with activation key: %w", usr.Name, err)
 	}
 	return activationKey, nil
 }
 
 // RemoveUser marks the user as removed and removes any credentials or
 // activation codes for the current users. Once a user is removed they are no
 // longer usable in Juju and should never be un removed.
-//
-// The following error types are possible from this function:
-// - usererrors.UsernameNotValid: When the username supplied is not valid.
-// - usererrors.NotFound: If no user by the given name exists.
-func (s *Service) RemoveUser(ctx context.Context, name string) error {
-	if err := ValidateUsername(name); err != nil {
-		return fmt.Errorf("username %q: %w", name, err)
+func (s *Service) RemoveUser(ctx context.Context, uuid user.UUID) error {
+	if err := uuid.Validate(); err != nil {
+		return fmt.Errorf("removing user uuid %q: %w", uuid, err)
```

Comment:
should this be "removing user for uuid"? I think it is more inline with what is going on as we are removing a user not a uuid.